File: ssm-sdk-client/src/main/java/ssm/client/SsmClientBuilder.java
package ssm.client;

import java.io.IOException;

import com.google.common.base.Preconditions;

import ssm.client.json.JSONConverter;
import ssm.client.json.ObjectMapperJSONConverter;
import ssm.client.repository.CoopRepository;
import ssm.client.repository.RepositoryFactory;

public class SsmClientBuilder {

    private CoopRepository coopRepository;
    private JSONConverter jsonConverter;

    private String channelId;
    private String ssmId;

    public static SsmClientBuilder builder() {
        return new SsmClientBuilder();
    }

    public static SsmClientBuilder builder(String filename) throws IOException {
        SsmClientConfig config = SsmClientConfig.fromConfigFile(filename);
        return builder(config);
    }


    public static SsmClientBuilder builder(SsmClientConfig config) {
        RepositoryFactory factory = new RepositoryFactory(config.getBaseUrl(), config.getBearerToken());
        CoopRepository coopRepository = factory.buildCoopRepository();
        JSONConverter converter = new ObjectMapperJSONConverter();
        return SsmClientBuilder.builder()
                .withCoopRepository(coopRepository)
                .withJSONConverter(converter)
                .withChannelId(config.getChannelId())
                .withSsmId(config.getChaincodeId());
    }



    public SsmClientBuilder() {
    }

    public SsmClientBuilder withCoopRepository(CoopRepository coopRepository) {
        this.coopRepository = coopRepository;
        return this;
    }

    public SsmClientBuilder withJSONConverter(JSONConverter jsonConverter) {
        this.jsonConverter = jsonConverter;
        return this;
    }

    public SsmClientBuilder withChannelId(String channelId) {
        this.channelId = channelId;
        return this;
    }

    public SsmClientBuilder withSsmId(String ssmId) {
        this.ssmId = ssmId;
        return this;
    }

    public SsmClient build() {
        Preconditions.checkNotNull(coopRepository);
        Preconditions.checkNotNull(jsonConverter);
        return new SsmClient(new SsmRequester(channelId, ssmId, jsonConverter, coopRepository));
    }
}

File: ssm-sdk/ssm-sdk-core/src/main/kotlin/ssm/chaincode/client/SsmClientBuilder.kt
package ssm.chaincode.client

import java.io.IOException
import ssm.chaincode.client.repository.KtorRepository
//import ssm.chaincode.client.repository.CoopRepository
//import ssm.chaincode.client.repository.RepositoryFactory
import ssm.sdk.json.JSONConverter
import ssm.sdk.json.JSONConverterObjectMapper

class SsmClientBuilder(
	private var coopRepository: KtorRepository ? = null,
	private var jsonConverter: JSONConverter? = null,
	private var channelId: String? = null,
	private var ssmId: String? = null,
) {
	fun withCoopRepository(coopRepository: KtorRepository?): SsmClientBuilder {
		this.coopRepository = coopRepository
		return this
	}

	fun withJSONConverter(jsonConverter: JSONConverter?): SsmClientBuilder {
		this.jsonConverter = jsonConverter
		return this
	}

	fun withChannelId(channelId: String?): SsmClientBuilder {
		this.channelId = channelId
		return this
	}

	fun withSsmId(ssmId: String?): SsmClientBuilder {
		this.ssmId = ssmId
		return this
	}

	fun build(): SsmClient {
		return SsmClient(SsmRequester(channelId, ssmId, jsonConverter!!, coopRepository!!))
	}

	companion object {
		fun builder(): SsmClientBuilder {
			return SsmClientBuilder()
		}

		@Throws(IOException::class)
		fun builder(filename: String): SsmClientBuilder {
			val config = SsmClientConfig.fromConfigFile(filename)
			return builder(config)
		}

		fun builder(config: SsmClientConfig): SsmClientBuilder {
			val coopRepository = KtorRepository(config.baseUrl, config.bearerToken)
			val converter: JSONConverter = JSONConverterObjectMapper()
			return builder()
				.withCoopRepository(coopRepository)
				.withJSONConverter(converter)
		}
	}
}